File: src/Entity/Utilisateur.php
<?php

namespace App\Entity;
use App\Entity\Planification;
use Symfony\Component\HttpFoundation\File\File;
use App\Entity\RendezVous;
use App\Enum\enumRole;
use App\Repository\UtilisateurRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Security\Core\User\UserInterface;
use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;
use Symfony\Component\Validator\Constraints as Assert; 

#[ORM\Entity(repositoryClass: UtilisateurRepository::class)]
class Utilisateur  implements UserInterface, PasswordAuthenticatedUserInterface
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "Le nom est nécessaire")] 
    #[Assert\Length(min:4,minMessage: "Veuillez avoir au moin 4 caractères")] 
    private ?string $nom = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "Le prenom est nécessaire")] 
    #[Assert\Length(min:4,minMessage: "Veuillez avoir au moin 4 caractères")] 
    private ?string $prenom = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "L'email est nécessaire")]
    #[Assert\Email(message: 'email {{value}} is not a valid email')] 
    private ?string $email = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "Le mot de passe est nécessaire")] 
    #[Assert\Length(min:8,minMessage:"Votre mot  de passe ne contient pas {{ limit }} caractères.")]
    private ?string $password = null;

    #[ORM\Column(type: Types::DATE_MUTABLE)]
    #[Assert\NotBlank(message: "La date de naissance est requise.")]
    #[Assert\Type(\DateTimeInterface::class, message: "Veuillez entrer une date valide.")]
    #[Assert\LessThan("today", message: "La date de naissance doit être dans le passé.")]
    private ?\DateTimeInterface $date_naissance;

    #[ORM\Column(type: "string", enumType: enumRole::class)]
    private enumRole $role;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "L'adresse est nécessaire")]
    #[Assert\NotNull(message: "L'adresse ne peut pas être nulle")]
    #[Assert\Length(min: 5, minMessage: "L'adresse doit contenir au moins {{ limit }} caractères")]
    private ?string $adresse = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "Le genre est nécessaire")]
    private ?string $genre = null;

    #[ORM\Column(length: 255)]
    private ?string $image = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "Le numero de telephone est nécessaire")]
    private ?string $numTel ;
    
    #[ORM\Column(type: 'boolean')]
    private $is_verified = false;


    #[ORM\Column(type: 'string',length:100,nullable: true)]
    private $resetToken;

    #[ORM\Column(type: 'boolean')]
    private $Tel_verified = false;

    #[ORM\Column(type: 'string',length:100,nullable: true)]
    private $codeVerif;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "L'image est nécessaire")]
    private ?string $image1 = null;
    /**
     * @var Collection<int, RendezVous>
     */
    #[ORM\OneToMany(mappedBy: 'id_medecin', targetEntity: RendezVous::class)]
    private Collection $rendezVouses;

    /**
     * @var Collection<int, RendezVous>
     */
    #[ORM\OneToMany(mappedBy: 'id_patient', targetEntity: RendezVous::class)]
    private Collection $rendez_vous;

    /**
     * @var Collection<int, Article>
     */
    #[ORM\OneToMany(mappedBy: 'id_medecin', targetEntity: Article::class)]
    private Collection $article;

    /**
     * @var Collection<int, Commentaire>
     */
    #[ORM\OneToMany(mappedBy: 'id_utilisateur', targetEntity: Commentaire::class)]
    private Collection $commentaire;


   

    #[ORM\OneToOne(targetEntity: Matching::class, mappedBy: 'utilisateur', cascade: ['persist', 'remove'])]
    private ?Matching $matching = null;

    private array $roles = [];

    /**
     * @var Collection<int, Planification>
     */
    #[ORM\OneToMany(targetEntity: Planification::class, mappedBy: 'id_freelancer')]
    private Collection $planifications;

    /**
     * @var Collection<int, ArticleRating>
     */
    #[ORM\OneToMany(mappedBy: 'user', targetEntity: ArticleRating::class)]
    private Collection $articleRatings;
    public function __construct()
    {
        $this->rendezVouses = new ArrayCollection();
        $this->rendez_vous = new ArrayCollection();
        $this->article = new ArrayCollection();
        $this->commentaire = new ArrayCollection();
        $this->articleRatings = new ArrayCollection();
        $this->planifications = new ArrayCollection();
        
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getNom(): ?string
    {
        return $this->nom;
    }

    public function setNom(string $nom): static
    {
        $this->nom = $nom;

        return $this;
    }

    public function getPrenom(): ?string
    {
        return $this->prenom;
    }

    public function setPrenom(string $prenom): static
    {
        $this->prenom = $prenom;

        return $this;
    }

    public function getEmail(): ?string
    {
        return $this->email;
    }

    public function setEmail(string $email): static
    {
        $this->email = $email;

        return $this;
    }

    public function getPassword(): ?string
    {
        return $this->password;
    }

    public function setPassword(string $password): static
    {
        $this->password = $password;

        return $this;
    }

    public function getDateNaissance(): ?\DateTimeInterface
    {
        return $this->date_naissance;
    }

    public function setDateNaissance(\DateTimeInterface $date_naissance): static
    {
        $this->date_naissance = $date_naissance;

        return $this;
    }

    public function getRole(): ?enumRole
    {
        return $this->role;
    }

    public function setRole(enumRole $role): static
    {
        $this->role = $role;

        return $this;
    }

    public function getAdresse(): ?string
    {
        return $this->adresse;
    }

    public function setAdresse(string $adresse): static
    {
        $this->adresse = $adresse;

        return $this;
    }

    public function getGenre(): ?string
    {
        return $this->genre;
    }

    public function setGenre(string $genre): static
    {
        $this->genre = $genre;

        return $this;
    }
    public function getImage(): ?string
    {
        return $this->image;
    }

    public function setImage(string $image): static
    {
        $this->image = $image;

        return $this;
    }

    public function getNumTel(): ?string
    {
        return $this->numTel;
    }

    public function setNumTel(string $numTel): static
    {
        $this->numTel = $numTel;

        return $this;
    }

    public function getIsVerified(): ?bool
    {
        return $this->is_verified;
    }

    public function setIsVerified(bool $is_verified): self
    {
        $this->is_verified = $is_verified;

        return $this;
    }

    public function getTelVerified(): ?bool
    {
        return $this->Tel_verified;
    }

    public function setTelVerified(bool $Tel_verified): self
    {
        $this->Tel_verified = $Tel_verified;

        return $this;
    }

    public function getResetToken (): ?string
    {
        return $this->resetToken;
    }

    public function setResetToken (string $resetToken): self
    {
        $this->resetToken = $resetToken;

        return $this;
    }

    public function getCodeVerif (): ?string
    {
        return $this->codeVerif;
    }

    public function setCodeVerif (string $codeVerif): self
    {
        $this->codeVerif = $codeVerif;

        return $this;
    }

    public function getImage1(): ?string
    {
        return $this->image1;
    }

    public function setImage1(string $image1): static
    {
        $this->image1 = $image1;

        return $this;
    }

    /**
     * @return Collection<int, RendezVous>
     */
    public function getRendezVouses(): Collection
    {
        return $this->rendezVouses;
    }

    public function addRendezVouse(RendezVous $rendezVouse): static
    {
        if (!$this->rendezVouses->contains($rendezVouse)) {
            $this->rendezVouses->add($rendezVouse);
            $rendezVouse->setIdMedecin($this);
        }

        return $this;
    }

    public function removeRendezVouse(RendezVous $rendezVouse): static
    {
        if ($this->rendezVouses->removeElement($rendezVouse)) {
            // set the owning side to null (unless already changed)
            if ($rendezVouse->getIdMedecin() === $this) {
                $rendezVouse->setIdMedecin(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, RendezVous>
     */
    public function getRendezVous(): Collection
    {
        return $this->rendez_vous;
    }

    public function addRendezVou(RendezVous $rendezVou): static
    {
        if (!$this->rendez_vous->contains($rendezVou)) {
            $this->rendez_vous->add($rendezVou);
            $rendezVou->setIdPatient($this);
        }

        return $this;
    }

    public function removeRendezVou(RendezVous $rendezVou): static
    {
        if ($this->rendez_vous->removeElement($rendezVou)) {
            // set the owning side to null (unless already changed)
            if ($rendezVou->getIdPatient() === $this) {
                $rendezVou->setIdPatient(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, Article>
     */
    public function getArticle(): Collection
    {
        return $this->article;
    }



    /**
     * @return Collection<int, Commentaire>
     */
    public function getCommentaire(): Collection
    {
        return $this->commentaire;
    }

    public function addCommentaire(Commentaire $commentaire): static
    {
        if (!$this->commentaire->contains($commentaire)) {
            $this->commentaire->add($commentaire);
            $commentaire->setIdUtilisateur($this);
        }

        return $this;
    }

    public function removeCommentaire(Commentaire $commentaire): static
    {
        if ($this->commentaire->removeElement($commentaire)) {
            // set the owning side to null (unless already changed)
            if ($commentaire->getIdUtilisateur() === $this) {
                $commentaire->setIdUtilisateur(null);
            }
        }

        return $this;
    }

    
    public function eraseCredentials(): void
    {
        // Effacer les informations sensibles, si nécessaire
    }

    public function getUserIdentifier(): string
    {
        return $this->email;
    }

    public function getRoles(): array
    {
        return [strtoupper($this->role->value)];
    }
    
    public function getSalt(): ?string
    {
        return null; // Pas nécessaire si vous utilisez bcrypt ou sodium
    }
   

    public function getMatching(): ?Matching
    {
        return $this->matching;
    }

    public function setMatching(?Matching $matching): static
    {
        // Ensure bidirectional relationship is set correctly
        if ($matching && $matching->getUtilisateur() !== $this) {
            $matching->setUtilisateur($this);
        }

        $this->matching = $matching;
        return $this;
    }
    /**
     * @return Collection<int, Planification>
     */
    public function getPlanifications(): Collection
    {
        return $this->planifications;
    }

    public function addPlanification(Planification $planification): static
    {
        if (!$this->planifications->contains($planification)) {
            $this->planifications->add($planification);
            $planification->setFreelancer($this);
        }
        return $this;
    }

    public function removePlanification(Planification $planification): static
    {
        if ($this->planifications->removeElement($planification)) {
            if ($planification->getFreelancer() === $this) {
                $planification->setFreelancer(null);
            }
        }
        return $this;
    }

    /**
     * @return Collection<int, ArticleRating>
     */
    public function getArticleRatings(): Collection
    {
        return $this->articleRatings;
    }

    public function addArticleRating(ArticleRating $articleRating): static
    {
        if (!$this->articleRatings->contains($articleRating)) {
            $this->articleRatings->add($articleRating);
            $articleRating->setUser($this);
        }

        return $this;
    }

    public function removeArticleRating(ArticleRating $articleRating): static
    {
        if ($this->articleRatings->removeElement($articleRating)) {
            // set the owning side to null (unless already changed)
            if ($articleRating->getUser() === $this) {
                $articleRating->setUser(null);
            }
        }

        return $this;
    }
    
    public function __toString(): string
    {
        return $this->nom . ' ' . $this->prenom; // Or any meaningful representation
    }
    

}

File: src/main/java/org/example/entities/Utilisateur.java
package org.example.entities;
import java.time.LocalDate;
import java.util.Date;
import java.util.Locale;

public class Utilisateur {
    private Matching matching;

    private int id;
    private String nom = null;
    private String prenom = null;
    private String email = null;
    private String password = null;
    private LocalDate date_naissance;

    private EnumRole role;
    private String adresse = null;
    private String genre = null;
   private String image = null;
    private String numTel ;
    private boolean is_verified = false;
    private String resetToken;
    private boolean Tel_verified = false;
    private Status status;
    private String image1;
    public Utilisateur(){}

    @Override
    public String toString() {
        return "Utilisateur{" +
                "id=" + id +
                ", nom='" + nom + '\'' +
                ", prenom='" + prenom + '\'' +
                ", email='" + email + '\'' +
                ", password='" + password + '\'' +
                ", date_naissance=" + date_naissance +
                ", role=" + role +
                ", adresse='" + adresse + '\'' +
                ", genre='" + genre + '\'' +
                ", numTel='" + numTel + '\'' +
                ", image='" + image1 + '\'' +
                '}';
    }

    public Utilisateur(String nom, String prenom, String email, String password, LocalDate date_naissance, EnumRole role, String adresse, String genre, String numTel,String image1) {
        this.nom = nom;
        this.prenom = prenom;
        this.email = email;
        this.role=role;
        this.date_naissance=date_naissance;
        this.password = password;
        this.adresse = adresse;
        this.genre = genre;
        this.numTel = numTel;
        this.image1=image1;
    }
    public Utilisateur(int id, String nom, String prenom, String email,EnumRole role,LocalDate date_naissance, String password, String adresse, String genre, String numTel) {
        this.id = id;
        this.nom = nom;
        this.prenom = prenom;
        this.email = email;
        this.role=role;
        this.date_naissance=date_naissance;
        this.password = password;
        this.adresse = adresse;
        this.genre = genre;
        this.numTel = numTel;
    }
    public Utilisateur(int id, String nom, String prenom, String email,EnumRole role,LocalDate date_naissance, String password, String adresse, String genre, String image, String numTel, boolean is_verified, String resetToken, boolean tel_verified, Status status, String image1) {
        this.id = id;
        this.nom = nom;
        this.prenom = prenom;
        this.email = email;
        this.role=role;
        this.date_naissance=date_naissance;
        this.password = password;
        this.adresse = adresse;
        this.genre = genre;
        this.image = image;
        this.numTel = numTel;
        this.is_verified = is_verified;
        this.resetToken = resetToken;
        Tel_verified = tel_verified;
        this.status = status;
        this.image1 = image1;
    }
    public Utilisateur(String nom, String prenom, String email,String password,LocalDate date_naissance,EnumRole role, String adresse, String genre, String image, String numTel, boolean is_verified, String resetToken, boolean tel_verified, Status status, String image1) {
        this.nom = nom;
        this.prenom = prenom;
        this.email = email;
        this.role=role;
        this.date_naissance=date_naissance;
        this.password = password;
        this.adresse = adresse;
        this.genre = genre;
        this.image = image;
        this.numTel = numTel;
        this.is_verified = is_verified;
        this.resetToken = resetToken;
        Tel_verified = tel_verified;
        this.status = status;
        this.image1 = image1;
    }


    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNom() {
        return nom;
    }

    public void setNom(String nom) {
        this.nom = nom;
    }

    public String getPrenom() {
        return prenom;
    }

    public void setPrenom(String prenom) {
        this.prenom = prenom;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }
    public LocalDate getDate_naissance() {
        return date_naissance;
    }

    public void setDate_naissance(LocalDate date_naissance) {
        this.date_naissance = date_naissance;
    }
    public EnumRole getRole() {
        return role;
    }

    public void setRole(EnumRole role) {
        this.role = role;
    }
    public String getAdresse() {
        return adresse;
    }

    public void setAdresse(String adresse) {
        this.adresse = adresse;
    }

    public String getGenre() {
        return genre;
    }

    public void setGenre(String genre) {
        this.genre = genre;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public String getnumTel() {
        return numTel;
    }

    public void setnumTel(String numTel) {
        this.numTel = numTel;
    }

    public boolean isIs_verified() {
        return is_verified;
    }

    public void setIs_verified(boolean is_verified) {
        this.is_verified = is_verified;
    }

    public String getResetToken() {
        return resetToken;
    }

    public void setResetToken(String resetToken) {
        this.resetToken = resetToken;
    }

    public boolean isTel_verified() {
        return Tel_verified;
    }

    public void setTel_verified(boolean tel_verified) {
        Tel_verified = tel_verified;
    }

    public Status getStatus() {
        return status;
    }

    public void setStatus(Status status) {
        this.status = status;
    }

    public String getImage1() {
        return image1;
    }

    public void setImage1(String image1) {
        this.image1 = image1;
    }

    public Matching getMatching() {
        return matching;
    }

    public void setMatching(Matching matching) {
        this.matching = matching;
    }

}
